File: sdk/java/src/main/java/io/agentsafe/spl/Node.java
package io.agentsafe.spl;

import java.util.List;
import java.util.Objects;

/**
 * AST node for SPL S-expressions. Sealed interface with record variants.
 */
public sealed interface Node {
    record BoolNode(boolean value) implements Node {}
    record NumNode(double value) implements Node {}
    record StrNode(String value) implements Node {}
    record SymNode(String name) implements Node {}
    record ListNode(List<Node> items) implements Node {}
    record NilNode() implements Node {}

    static Node bool_(boolean b) { return new BoolNode(b); }
    static Node num(double n) { return new NumNode(n); }
    static Node str(String s) { return new StrNode(s); }
    static Node sym(String s) { return new SymNode(s); }
    static Node list(List<Node> items) { return new ListNode(items); }
    static Node nil() { return new NilNode(); }

    default boolean isTruthy() {
        return switch (this) {
            case BoolNode b -> b.value();
            case NilNode n -> false;
            case NumNode n -> n.value() != 0.0;
            default -> true;
        };
    }

    default double asDouble() {
        return switch (this) {
            case NumNode n -> n.value();
            default -> 0.0;
        };
    }

    default String asString() {
        return switch (this) {
            case StrNode s -> s.value();
            case SymNode s -> s.name();
            case BoolNode b -> String.valueOf(b.value());
            case NumNode n -> String.valueOf(n.value());
            case NilNode ignored -> "nil";
            case ListNode l -> l.toString();
        };
    }

    static boolean nodeEq(Node a, Node b) {
        if (a instanceof StrNode s1 && b instanceof StrNode s2) return s1.value().equals(s2.value());
        if (a instanceof SymNode s1 && b instanceof SymNode s2) return s1.name().equals(s2.name());
        if (a instanceof StrNode s && b instanceof SymNode y) return s.value().equals(y.name());
        if (a instanceof SymNode s && b instanceof StrNode y) return s.name().equals(y.value());
        if (a instanceof NumNode n1 && b instanceof NumNode n2) return n1.value() == n2.value();
        if (a instanceof BoolNode b1 && b instanceof BoolNode b2) return b1.value() == b2.value();
        if (a instanceof NilNode && b instanceof NilNode) return true;
        return a.asString().equals(b.asString());
    }
}

File: sdk/csharp/AgentSafe.Spl/Node.cs
namespace AgentSafe.Spl;

/// <summary>AST node for SPL S-expressions.</summary>
public abstract record Node
{
    public sealed record BoolNode(bool Value) : Node;
    public sealed record NumNode(double Value) : Node;
    public sealed record StrNode(string Value) : Node;
    public sealed record SymNode(string Name) : Node;
    public sealed record ListNode(List<Node> Items) : Node;
    public sealed record NilNode() : Node;

    public static Node Bool(bool b) => new BoolNode(b);
    public static Node Num(double n) => new NumNode(n);
    public static Node Str(string s) => new StrNode(s);
    public static Node Sym(string s) => new SymNode(s);
    public static Node List(List<Node> items) => new ListNode(items);
    public static Node Nil() => new NilNode();

    public bool IsTruthy() => this switch
    {
        BoolNode b => b.Value,
        NilNode => false,
        NumNode n => n.Value != 0.0,
        _ => true
    };

    public double AsDouble() => this switch
    {
        NumNode n => n.Value,
        _ => 0.0
    };

    public string AsString() => this switch
    {
        StrNode s => s.Value,
        SymNode s => s.Name,
        BoolNode b => b.Value.ToString().ToLower(),
        NumNode n => n.Value.ToString(),
        NilNode => "nil",
        ListNode l => $"({string.Join(" ", l.Items)})",
        _ => ""
    };

    public static bool NodeEq(Node a, Node b)
    {
        if (a is StrNode sa && b is StrNode sb) return sa.Value == sb.Value;
        if (a is SymNode ya && b is SymNode yb) return ya.Name == yb.Name;
        if (a is StrNode s1 && b is SymNode y1) return s1.Value == y1.Name;
        if (a is SymNode y2 && b is StrNode s2) return y2.Name == s2.Value;
        if (a is NumNode na && b is NumNode nb) return na.Value == nb.Value;
        if (a is BoolNode ba && b is BoolNode bb) return ba.Value == bb.Value;
        if (a is NilNode && b is NilNode) return true;
        return a.AsString() == b.AsString();
    }
}
